AppCenter/AppCenter/Internals/Util/MSKeychainUtil.m

```diff
@@ -17,6 +17,33 @@ @implementation MSKeychainUtil
   return serviceName;
 }
 
++ (BOOL)storeArray:(NSArray *)array forKey:(NSString *)key {
```

Comment:
I don't think we should store arrays of objects in the keychain. One token is OK but for multiple tokens I would do something similar to what we are already doing for 1DS tokens: Use the MSEncrypter to encrypt the token and store it. 1DS tokens are stored with logs in the DB but auth tokens history can be stored in the user defaults with the actual token being encrypted by the encrypter of course.

Reply:
I just wonder if there are any limitations on the keychain item size or if there are some performance issues. Currently, we're limiting the history by just 5 items.

Reply:
@clpolet Disagree.

This is a list of tokens that can still be valid. Also, additional information is just some integers.
Plus we have a size limit of history and cache decrypted value. I believe that's not an issue.

Reply:
@jaelim-ms for now, yes, we can refactor this with further pull requests. This one is already big enough and has too many changes :(